rust/chains/hyperlane-ethereum/src/merkle_tree_hook.rs

```diff
@@ -0,0 +1,282 @@
+        &self,
+        provider: M,
+        locator: &ContractLocator,
+    ) -> Self::Output {
+        Box::new(EthereumMerkleTreeHookIndexer::new(
+            Arc::new(provider),
+            locator,
+            self.finality_blocks,
+        ))
+    }
+}
+
+#[derive(Debug)]
+/// Struct that retrieves event data for an Ethereum MerkleTreeHook
+pub struct EthereumMerkleTreeHookIndexer<M>
+where
+    M: Middleware,
+{
+    contract: Arc<MerkleTreeHookContract<M>>,
+    provider: Arc<M>,
+    finality_blocks: u32,
+}
+
+impl<M> EthereumMerkleTreeHookIndexer<M>
+where
+    M: Middleware + 'static,
+{
+    /// Create new EthereumMerkleTreeHookIndexer
+    pub fn new(provider: Arc<M>, locator: &ContractLocator, finality_blocks: u32) -> Self {
+        Self {
+            contract: Arc::new(MerkleTreeHookContract::new(
+                locator.address,
+                provider.clone(),
+            )),
+            provider,
+            finality_blocks,
+        }
+    }
+}
+
+#[async_trait]
+impl<M> Indexer<MerkleTreeInsertion> for EthereumMerkleTreeHookIndexer<M>
+where
+    M: Middleware + 'static,
+{
+    #[instrument(err, skip(self))]
+    async fn fetch_logs(
+        &self,
+        range: RangeInclusive<u32>,
+    ) -> ChainResult<Vec<(MerkleTreeInsertion, LogMeta)>> {
+        let events = self
+            .contract
+            .inserted_into_tree_filter()
+            .from_block(*range.start())
+            .to_block(*range.end())
+            .query_with_meta()
+            .await?;
+
+        let logs = events
+            .into_iter()
+            .map(|(log, log_meta)| {
+                (
+                    MerkleTreeInsertion::new(log.index, H256::from(log.message_id)),
+                    log_meta.into(),
+                )
+            })
+            .collect();
+        Ok(logs)
+    }
+
+    #[instrument(level = "debug", err, ret, skip(self))]
+    async fn get_finalized_block_number(&self) -> ChainResult<u32> {
+        Ok(self
+            .provider
+            .get_block_number()
+            .await
+            .map_err(ChainCommunicationError::from_other)?
+            .as_u32()
+            .saturating_sub(self.finality_blocks))
+    }
+}
+
+#[async_trait]
+impl<M> SequenceIndexer<MerkleTreeInsertion> for EthereumMerkleTreeHookIndexer<M>
+where
+    M: Middleware + 'static,
+{
+    async fn sequence_and_tip(&self) -> ChainResult<(Option<u32>, u32)> {
+        // The InterchainGasPaymasterIndexerBuilder must return a `SequenceIndexer` type.
+        // It's fine if only a blanket implementation is provided for EVM chains, since their
+        // indexing only uses the `Index` trait, which is a supertrait of `SequenceIndexer`.
+        // TODO: if `SequenceIndexer` turns out to not depend on `Indexer` at all, then the supertrait
+        // dependency could be removed, even if the builder would still need to return a type that is both
+        // ``SequenceIndexer` and `Indexer`.
```

Comment:
should we be tracking this somewhere? I'm not sure I understand all the implications @tkporter

Reply:
if we do get around to revamping indexing, this will certainly be addressed one way or another. But I think it's low prio for now

Reply:
+1 this 100% needs refactoring at some point but it is a big undertaking

I'd probably put this in the realm of https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/2313